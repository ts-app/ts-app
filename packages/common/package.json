{
  "name": "@ts-app/common",
  "version": "0.1.3-2",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "license": "MIT",
  "repository": "https://github.com/ts-app/ts-app/tree/master/packages/common",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "clean": "rm -fr ./dist",
    "compile": "tsc -p config/tsconfig-compile.json",
    "compile:watch": "tsc -p config/tsconfig-compile.json --watch",
    "develop": "npm run compile:watch",
    "doc": "typedoc --mode file --module commonjs --target ES6 --out dist/docs src",
    "lint": "tslint --project config/tsconfig-test.json",
    "prepublishOnly": "npm run clean && npm run lint && npm run compile && npm test",
    "test": "jest --config config/jest.config.js",
    "test:coverage": "jest --config config/jest.config.js --coverage --collectCoverage",
    "test:watch": "jest --config config/jest.config.js --watch"
  },
  "dependencies": {
    "lz-string": "^1.4.4",
    "rxjs": "^6.2.1"
  },
  "devDependencies": {
    "@types/lz-string": "^1.3.32",
    "jest": "^23.1.0",
    "tslint": "^5.10.0",
    "typedoc": "^0.11.1",
    "typescript": "^2.9.2"
  }
}
