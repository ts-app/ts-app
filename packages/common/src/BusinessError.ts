/**
 * This interface represents recoverable errors.
 */
export interface BusinessError {
  /**
   * Business error message.
   */
  message: string
}
