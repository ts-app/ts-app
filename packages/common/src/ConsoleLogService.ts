import { LogService } from './LogService'

/**
 * LogService implementation that writes messages to console.
 */
export class ConsoleLogService extends LogService {
  error (message: string, error?: any): void {
    if (error) {
      console.error(message)
    } else {
      console.error(message, error)
    }
  }

  log (message: string): void {
    console.log(message)
  }

  debug (message: string | object): void {
    console.debug(message)
  }
}
