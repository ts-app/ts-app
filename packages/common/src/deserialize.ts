import * as LZString from 'lz-string'

export function deserialize (data: any, options?: {
  mode:
    'direct' |
    'decompress' |
    'decompressToUTF16' |
    'decompressToBase64' |
    'decompressToEncodedURIComponent'
}) {
  const reviver = (key: any, data: any) => {
    if (typeof data === 'string' && /^\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d.\d\d\dZ$/.test(data)) {
      return new Date(data)
    }
    return data
  }

  const mode = options && options.mode || 'direct'
  let val: any

  switch (mode) {
    case 'direct':
      val = data
      break
    case 'decompress':
      val = LZString.decompress(data)
      break
    case 'decompressToUTF16':
      val = LZString.decompressFromUTF16(data)
      break
    case 'decompressToBase64':
      val = LZString.decompressFromBase64(data)
      break
    case 'decompressToEncodedURIComponent':
      val = LZString.decompressFromEncodedURIComponent(data)
      break
    default:
      throw new Error(`Unknown deserialization mode [${mode}]`)
  }

  return JSON.parse(val, reviver)
}
