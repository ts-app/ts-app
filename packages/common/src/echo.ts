export const echo = (msg: string) => (`>>> ${msg}`)
