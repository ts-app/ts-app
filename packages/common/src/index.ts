export * from './assert'
export * from './BusinessError'
export * from './ConsoleLogService'
export * from './echo'
export * from './escapeRegex'
export * from './FindInput'
export * from './FindOutput'
export * from './LogService'
export * from './makeTimestampable'
export * from './ofBusinessError'
export * from './omitVolatile'
export * from './Timestampable'
export * from './User'
export * from './UserProfile'
export * from './validateEmail'
