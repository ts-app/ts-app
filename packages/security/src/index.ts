export * from './MongoRoleService'
export * from './MongoSecurityService'
export * from './RoleService'
export * from './SecurityService'
