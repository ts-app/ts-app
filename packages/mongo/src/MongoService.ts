import { FindInput, FindOutput, LogService } from '@ts-app/common'
import { Observable, from, of, throwError } from 'rxjs'
import {
  Collection, Cursor, Db, DeleteWriteOpResultObject, FindOneOptions, MongoClient, ObjectId
} from 'mongodb'
import { concatMap, map, mapTo } from 'rxjs/operators'
import { escapeRegex } from '../../common/dist'

/**
 * Returns an object where read/write to "id" property is mapped to Mongo's Object ID.
 */
const toMongoDoc = (doc: any) => {
  if (doc._id !== undefined) {
    return {
      ...doc,
      get id () {
        return doc._id ? doc._id.toString() : undefined
      },
      set id (id: string) {
        this._id = new ObjectId(id)
      }
    }
  } else {
    return doc
  }
}

/**
 * MongoService provides to access MongoDB via promise based functions for common usage patterns such as CRUD and pagination via cursor.
 */
export class MongoService {
  private _db?: Db
  private client!: MongoClient

  constructor (private mongoUrl: string,
               private logService?: LogService) {
  }

  create<T> (collectionName: string, doc: Pick<T, Exclude<keyof T, 'id'>>): Observable<string> {
    return this.collection(collectionName).pipe(
      concatMap(collection => collection.insertOne(doc)),
      concatMap(result => {
        if (result.result.n !== 1) {
          return throwError(`Error creating document in collection [${collectionName}]`)
        } else {
          return of(result.insertedId.toString())
        }
      })
    )
  }

  get<T> (collectionName: string, idOrFilter: string | object, options?: FindOneOptions): Observable<T | null> {
    return this.collection(collectionName).pipe(
      concatMap(collection => {
        // if idOrFilter is string, place it in an object with key of _id
        let filter: any
        if (typeof idOrFilter === 'string') {
          filter = { _id: new ObjectId(idOrFilter) }
        } else {
          filter = idOrFilter
        }

        return collection.findOne(filter, options)
      }),
      map(doc => doc ? toMongoDoc(doc) : null)
    )
  }

  cursor<T> (collectionName: string, query: object): Observable<Cursor<T>> {
    return this.collection(collectionName).pipe(
      map(collection => collection.find(query))
    )
  }

  update (collectionName: string, id: string, update: object): Observable<null> {
    return this.collection(collectionName).pipe(
      concatMap(collection => collection.updateOne({ _id: new ObjectId(id) }, update)),
      concatMap(result => {
        if (result.result.ok !== 1) {
          return throwError(`Error updating document [${id}] in collection [${collectionName}]`)
        } else {
          return of(null)
        }
      })
    )
  }

  remove (collectionName: string, idOrFilter: string | object): Observable<number> {
    if (!idOrFilter) {
      return throwError('Invalid ID or filter')
    }

    return this.collection(collectionName).pipe(
      concatMap(collection => {
        let deleteById = false
        let filter: any
        if (typeof idOrFilter === 'string') {
          filter = { _id: new ObjectId(idOrFilter) }
          deleteById = true
        } else {
          filter = idOrFilter
        }

        if (deleteById) {
          return collection.deleteOne(filter)
        } else {
          return collection.deleteMany(filter)
        }
      }),
      map((result: DeleteWriteOpResultObject | DeleteWriteOpResultObject) => {
        return result.result.n || 0
      })
    )
  }

  collection (collectionName: string): Observable<Collection> {
    return this.db().pipe(
      map((db: Db) => db.collection(collectionName))
    )
  }

  count (collectionName: string): Observable<number> {
    return this.collection(collectionName).pipe(
      concatMap(collection => collection.count())
    )
  }

  dropCollection (collectionName: string): Observable<boolean> {
    return this.collection(collectionName).pipe(
      concatMap((collection: Collection) => collection.drop())
    )
  }

  collectionExist (collectionName: string): Observable<boolean> {
    return this.db().pipe(
      concatMap(db => db.collections()),
      map(collections => !!collections.find(c => c.collectionName === collectionName))
    )
  }

  db (): Observable<Db> {
    if (!this._db) {
      // TODO: parameterize options
      return from(MongoClient.connect(this.mongoUrl)).pipe(
        map(client => this.updateClient(client))
      )
    } else {
      return of(this._db)
    }
  }

  close (): Observable<null> {
    if (this.client) {
      return from(this.client.close()).pipe(mapTo(null))
    } else {
      return of(null)
    }
  }

  findWithCursor<T> (
    collectionName: string, filter: object, limit: number = 10, cursor?: string,
    sort?: { field: string, asc: boolean }[], project?: object): Observable<FindOutput<T>> {
    return this.collection(collectionName).pipe(
      concatMap((collection: Collection) => {
        let filterWithCursor = filter
        if (cursor) {
          filterWithCursor = {
            ...filterWithCursor,
            _id: { $gt: new ObjectId(cursor) }
          }
        }

        let mongoCursor = collection
          .find(filterWithCursor)
          .limit(limit)

        // sort is parsed before passing to sort()
        if (sort) {
          const sortObject = sort.reduce((acc, current) => {
            acc[ current.field ] = current.asc ? 1 : -1
            return acc
          }, {} as any)
          mongoCursor = mongoCursor.sort(sortObject)
        }

        // project is passed directly to Mongo
        if (project) {
          mongoCursor = mongoCursor.project(project)
        }

        this.debug({
          filterWithCursor,
          sort,
          project,
          limit
        })

        return mongoCursor.toArray()
      }),
      map(docs => {
        const mongoDocs = docs.map(doc => toMongoDoc(doc))
        const lastDoc = mongoDocs.length > 0 ? mongoDocs[ mongoDocs.length - 1 ] : null

        return {
          cursor: lastDoc ? lastDoc.id : null,
          docs: mongoDocs
        }
      })
    )
  }

  find<T> (input: FindInput, collectionName: string, findBy: string[], defaultSort?: { field: string, asc: boolean }[]) {
    let filter = {}
    let { sort } = input
    const { q, limit, cursor, project } = input
    if (q && q.trim().length > 0) {
      filter = {
        $or: [
          ...[ findBy.map(name => {
            // why "input.q!"?
            // reason: https://github.com/cartant/rxjs-tslint-rules/issues/54
            return { [ name ]: { $regex: `^${escapeRegex(input.q!)}`, $options: 'i' } }
          }) ]
        ]
      }
    }

    // if sort is not specified in input, determine default sort
    // if default sort is not specified, derive from first findBy field
    if (!sort) {
      if (!defaultSort && findBy.length > 0) {
        defaultSort = [ { field: findBy[ 0 ], asc: true } ]
      }
      sort = defaultSort
    }

    return this.findWithCursor<T>(collectionName, filter, limit, cursor, sort, project)
  }

  private updateClient (client: MongoClient) {
    this.client = client
    this._db = client.db()
    return this._db
  }

  private debug (message: string | object) {
    if (this.logService) {
      this.logService.debug(message)
    }
  }
}
