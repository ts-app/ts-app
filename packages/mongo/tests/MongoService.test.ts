import { MongoService } from '../src'
import { catchError, concatMap, map, mapTo } from 'rxjs/operators'
import { of } from 'rxjs'
import { BusinessError } from '../../common/dist'

describe('MongoService', async () => {
  const localUrl = 'mongodb://localhost:27017'
  let mongoService: MongoService

  beforeEach(done => {
    mongoService = new MongoService(localUrl)
    mongoService.dropCollection('test').pipe(
      // just ignore if cannot drop
      catchError(() => {
        return of(null)
      })
    ).subscribe(() => done())
  })

  afterEach(done => mongoService.close().subscribe(() => done()))

  test('crud', done => {
    type DocSchema = {
      id: string
      title: string
      description: string
    }

    // --- create
    mongoService.create<DocSchema>('test', {
      id: '',
      title: 'test 123',
      description: 'test 123'
    }).pipe(
      map<{ error: BusinessError, id?: string }, string>(createResult => {
        expect(createResult.id!.length).toBe(24)
        return createResult.id!
      }),

      // --- read
      concatMap(id => mongoService.get<DocSchema>('test', id).pipe(
        map(doc => ({ doc, id }))
        )
      ),
      map(({ doc, id }) => {
        expect(doc!.id).toBe(id)
        expect(doc!.title).toBe('test 123')
        return id
      }),

      // --- update
      concatMap(id => {
        return mongoService.update('test', id, {
          $set: {
            title: 'changed 456'
          }
        }).pipe(mapTo(id))
      }),
      // TODO: assert update...
      concatMap((id) => mongoService.get<DocSchema>('test', id)),
      map(doc => {
        expect(doc!.title).toBe('changed 456')
        expect(doc!.description).toBe('test 123')
        return doc!.id
      }),

      // -- delete
      concatMap(id => mongoService.remove('test', id)),
      map(deleteResult => {
        expect(deleteResult).toBe(1)
      })
    ).subscribe(() => {
      expect.assertions(6)
      done()
    })
  })

  test('remove() invalid ID', done => {
    mongoService.remove('test', null as any).pipe(
      catchError(e => {
        expect(e).toEqual('Invalid ID or filter')
        return of(null)
      })
    )
      .subscribe(() => {
        expect.assertions(1)
        done()
      })
  })

  test('collectionExist()', done => {
    mongoService.dropCollection('test-collection').pipe(
      // ignore error, just proceed
      catchError(() => of(null)),
      concatMap(() => mongoService.collectionExist('test-collection')),
      map(exist => expect(exist).toBe(false)),
      concatMap(() => mongoService.create('test-collection', { sample: 123 })),
      concatMap(() => mongoService.collectionExist('test-collection'))
    ).subscribe(exist => {
      expect(exist).toBe(true)
      expect.assertions(2)
      done()
    })
  })

  test('get() with options', done => {
    type DocSchema = { name: string, description: string }

    mongoService.create<DocSchema>('test', { name: 'abc', description: 'def' }).pipe(
      // get all fields
      concatMap(() => mongoService.get<DocSchema>('test', { name: 'abc' })),
      map(get => {
        expect(get!.name).toBe('abc')
        expect(get!.description).toBe('def')
      }),

      // only projected fields
      concatMap(() => mongoService.get<DocSchema>('test',
        { name: 'abc' },
        {
          fields: {
            description: 1
          }
        }
      )),
      map(get => {
        expect(get!.name).toBeUndefined()
        expect(get!.description).toBe('def')
      })
    ).subscribe(() => {
      expect.assertions(4)
      done()
    })
  })

  test('create() with null', done => {
    mongoService.create('test', null)
      .subscribe(
        undefined,
        (e: TypeError) => {
          expect(e.message).toMatchSnapshot()
          expect.assertions(1)
          done()
        },
        () => {
          fail('Should not be completing!')
          done()
        }
      )
  })
})
