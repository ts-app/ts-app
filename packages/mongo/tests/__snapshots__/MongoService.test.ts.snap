// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoService get() with null 1`] = `"Cannot read property '_id' of null"`;
